packages/core/src/amazonqFeatureDev/util/upload.ts

```diff
@@ -30,8 +32,16 @@ export async function uploadCode(url: string, buffer: Buffer, checksumSha256: st
         }).response
     } catch (e: any) {
         getLogger().error(`${featureName}: failed to upload code to s3: ${(e as Error).message}`)
-        throw new UploadCodeError(
-            e instanceof RequestError ? `${e.response.status}: ${e.response.statusText}` : 'Unknown'
-        )
+        if (e instanceof RequestError) {
+            switch (e.response.status) {
+                case 403:
+                    throw new UploadURLExpired()
+                default:
+                    throw new UploadCodeError(
+                        e instanceof RequestError ? `${e.response.status}: ${e.response.statusText}` : 'Unknown'
+                    )
+            }
+        }
+        throw new ToolkitError(i18n('AWS.amazonq.featureDev.error.codeGen.default'))
```

Comment:
Try `ToolkitError.chain(e, i18n(...))` as it creates a new error, while wrapping the underlying error in the `cause` field. Then if this error reaches a telemetry call it will be "unwrapped" and include the chain of errors

Reply:
Will followup